### ISSUE TITLE
Random Output from yolov8 pose

### ISSUE THREAD

**Issue by akash-syook** [ultralytics/ultralytics#16035]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi Everyone,
I have trained a yolov8-m pose estimation model to detect custom keypoints. Following are my poses.
"head","left_foot","right_foot","left_hand","right_hand","left_knee","right_knee","up_body_1","up_body_2"

Adding below is my data.yaml file

train: output_directory/images
val:output_directory/images
test: output_directory/images

kpt_shape: [9, 3]
flip_idx: [0, 2, 1, 4, 3, 6, 5, 7, 8]

nc: 1
names: ['person']`

Sample annotations txt 
0 0.5 0.5 1.0 1.0 0.6515646353044727 0.2022792022792023 2 0.0 0.0 0 0.0 0.0 0 0.0 0.0 0 0.2737243184397656 0.8148148148148148 2 0.0 0.0 0 0.0 0.0 0 0.525617863016237 0.44017094017094016 2 0.42318115488847197 0.668091168091168 2

I have only kept the values which are visible and annotated (2). 
In above example "head", "right_hand","up_body_1","up_body_2"

I trained the model with below args
 yolo pose train data=data.yaml model=yolov8l-pose.pt epochs=200 imgsz=640  batch=4
 
 While predicting, I am  getting completely random outputs when testing on the training data itself.
 "val_batch1_labels.jpg" shows the correct positions of keypoints but "val_batch1_pred.jpg"  is complete random.
 
 Can anyone please point out what I'm doing wrong.

**Comment by Y-T-G** [comment]
Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by akash-syook** [comment]
yolo checks
Ultralytics YOLOv8.1.29 🚀 Python-3.10.13 torch-2.2.1 CUDA:0 (NVIDIA GeForce RTX 4070 Laptop GPU, 7843MiB)
Setup complete ✅ (16 CPUs, 15.3 GB RAM, 556.2/937.3 GB disk)

OS                  Linux-6.8.0-40-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.13
Install             pip
RAM                 15.33 GB
CPU                 12th Gen Intel Core(TM) i7-12650H
CUDA                11.8

matplotlib          ✅ 3.9.1>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.2.1>=1.8.0
torchvision         ✅ 0.17.1>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

**Comment by akash-syook** [comment]
Yes @Y-T-G I used the best.pt. Attached below the code.
from ultralytics import YOLO

# Load a model
# load an official model
model = YOLO("/runs/pose/train13/weights/best.pt")  # load a custom model

# Predict with the model
results = model("cropped_data/output_directory/images",conf=0.9,save=True,save_txt=True)  # predict on an image

**Comment by Y-T-G** [comment]
@akash-syook What's the validation score?

**Comment by Y-T-G** [comment]
You need to train longer. That's a really low mAP. 15 images means that's just 1 iteration per epoch. Or 100 iterations in 100 epochs. That's not enough for learning or convergence.

**Comment by akash-syook** [comment]
Well I thought about this too that's why I reduced the batch size to 1 and trained for 600 epochs.
Training stopped in between
`Stopping training early as no improvement observed in last 100 epochs. Best results observed at epoch 9, best model saved as best.pt.`
I made EarlyStopping(patience=300) but still the same. Epoch 9 gives the best result.

**Comment by glenn-jocher** [comment]
It seems like the model isn't improving beyond epoch 9, which could be due to overfitting or insufficient data. Consider increasing your dataset size or experimenting with data augmentation to enhance model performance.

**Comment by akash-syook** [comment]
Ok will try for 50 unique images and also try data augmentation to make it 150.
Will train and update you.

**Comment by Y-T-G** [comment]
`batch=1` would also only perform 1 weight update per epoch because there's gradient accumulation that occurs until the batch size is 64.

Also based on the loss plots, the model isn't learning well, and there's probably an issue with the labels or data.

Are you using the same set as `train` and `val`?

**Comment by akash-syook** [comment]
Data is not the issue from what I have seen beacuse;
"val_batch0_labels.jpg" and "val_batch1_labels.jpg" drawn shows correct points but 
"val_batch0_pred.jpg" and "val_batch1_pred.jpg" drawn are random. 
Yes I am using same data for `train` and `val`. I want to overfit the model on this data and if there are correct prediction, I can say my training pipeline is ready and the I will create a large amount of data for training.

**Comment by Y-T-G** [comment]
Are there symmetric points in your dataset? Did you correctly specify the `flip_idx`? Are all the labels following the the same order for keypoints? Do they correspond with actual visual features in the image?

**Comment by akash-syook** [comment]
`Are there symmetric points in your dataset? `: YES, eg:"left_knee","right_knee",
` Did you correctly specify the flip_idx?`: Yes as per my knowlegde
CLASSES:"head","left_foot","right_foot","left_hand","right_hand","left_knee","right_knee","up_body_1","up_body_2
kpt_shape: [9, 3]
flip_idx: [0, 2, 1, 4, 3, 6, 5, 7, 8]
Are all the labels following the the  same order for keypoints?  YEs,  if you mean  annotations of all image should be in order of 
CLASSES.
Do they correspond with actual visual features in the image?: I don't understand this point

**Comment by Y-T-G** [comment]
It seems the keypoints had the best improvement during the last epochs. So you can try training with `rect=True`.

**Comment by Y-T-G** [comment]
How many images do you have now?

Can you provide the output after running this command in terminal: `yolo checks`?

**Comment by akash-syook** [comment]
I have 70-80 images.
WARNING ⚠️ Ultralytics settings reset to default values. This may be due to a possible problem with your settings or a recent ultralytics package update. 
View settings with 'yolo settings' or at '/home/akash/.config/Ultralytics/settings.yaml'
Update settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics YOLOv8.2.90 🚀 Python-3.10.13 torch-2.2.1 CUDA:0 (NVIDIA GeForce RTX 4070 Laptop GPU, 7843MiB)
Setup complete ✅ (16 CPUs, 15.3 GB RAM, 806.2/937.3 GB disk)

OS                  Linux-6.8.0-45-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.13
Install             pip
RAM                 15.33 GB
CPU                 12th Gen Intel Core(TM) i7-12650H
CUDA                11.8

numpy               ✅ 1.24.3<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.1>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.14.0>=1.4.1
torch               ✅ 2.2.1>=1.8.0
torchvision         ✅ 0.17.1>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.6>=2.0.0
torch               ✅ 2.2.1!=2.4.0,>=1.8.0; sys_platform == "win32"

**Comment by Y-T-G** [comment]
Also post your training command and dataset yaml.

**Comment by Y-T-G** [comment]
It works with 17 keypoints, because the pretrained model also has 17 keypoints. So it just finetunes the existing head on your data.

If you have a different number of keypoints, then it has to create a new head for that number of keypoints which means the pretrained weights for the head can't be used and it has to start from randomized head. When it starts from a randomized head, it needs much more data to learn.

**Comment by akash-syook** [comment]
I understand this training data is not enough for production env but, it should atleast overfit on the training data. That's what I want to achieve to make sure we are doing all the steps correctly.
There is a problem with either framework or my dataset. 
Training command: yolo pose train data=data.yaml model=yolov8l-pose.pt epochs=3000 imgsz=640  batch=4
train: output_directory/images
val:output_directory/images
test: output_directory/images

kpt_shape: [9, 3]
flip_idx: [0, 2, 1, 4, 3, 6, 5, 7, 8]

nc: 1
names: ['person']`

**Comment by akash-syook** [comment]
sorry I'm not able to follow what are you suggesting.
Should I train from scratch using "yolo pose train data=data.yaml model=yolov8l-pose.yaml epochs=3000 imgsz=640  batch=1 patience=300"?

**Comment by Y-T-G** [comment]
Did you check the training batch plots? Does it show the keypoints correctly?

**Comment by Y-T-G** [comment]
Change visibility to be either 0 or 1.

In your sample annotation, you set the box as the whole image. Why?

The keypoints depend on correct bounding boxes.

**Comment by akash-syook** [comment]
Yes train_batch1.jpg,   train_batch2.jpg shows correct points 
In your sample annotation, you set the box as the whole image. Why?: Because I'm running the training on single person image.
Change visibility to be either 0 or 1.? Where?

**Comment by Y-T-G** [comment]
You set visibility to 2 in your sample annotation. If it's visible, it should be 1.

**Comment by Y-T-G** [comment]
> Because I'm running the training on single person image.

Then you should use a smaller imgsz (try 128) so that the box can cover the whole image.

**Comment by akash-syook** [comment]
https://github.com/ultralytics/ultralytics/issues/6945
This says the following:
0 indicates not visible and not labeled
1 indicates not visible but labeled
2 indicates visible and labeled
If my data is already in good quality, img size is 640 then I should use it right?
I have good quality videos

**Comment by Y-T-G** [comment]
Use 0 for invisible points and 1 for visible points. That's what is used in the pretrained models.

You shouldn't use 640 imgsz for single person images since the bounding box can't cover the whole image due to network receptive field limitation. It's also inappropriate usage of a detection model (YOLO-Pose performs detection too), since objects aren't supposed to be the whole image. Use smaller imgsz between 128 and 224.
